consensus/consensus-types/src/common.rs

```diff
@@ -233,18 +279,90 @@ impl Payload {
             },
         }
     }
 
-    pub fn extend(&mut self, other: Payload) {
+    pub fn extend(self, other: Payload) -> Self {
         match (self, other) {
-            (Payload::DirectMempool(v1), Payload::DirectMempool(v2)) => v1.extend(v2),
-            (Payload::InQuorumStore(p1), Payload::InQuorumStore(p2)) => p1.extend(p2),
+            (Payload::DirectMempool(v1), Payload::DirectMempool(v2)) => {
+                let mut v3 = v1;
+                v3.extend(v2);
+                Payload::DirectMempool(v3)
+            },
+            (Payload::InQuorumStore(p1), Payload::InQuorumStore(p2)) => {
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::InQuorumStore(p3)
+            },
             (Payload::InQuorumStoreWithLimit(p1), Payload::InQuorumStoreWithLimit(p2)) => {
-                p1.extend(p2)
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::InQuorumStoreWithLimit(p3)
+            },
+            (
+                Payload::QuorumStoreInlineHybrid(b1, p1, m1),
+                Payload::QuorumStoreInlineHybrid(b2, p2, m2),
+            ) => {
+                let mut b3 = b1;
+                b3.extend(b2);
+                let mut p3 = p1;
+                p3.extend(p2);
+                // TODO: What's the right logic here?
+                let m3 = if m1.is_none() {
+                    m2
+                } else if let Some(m2) = m2 {
+                    Some(m1.unwrap() + m2)
+                } else {
+                    m1
+                };
+                Payload::QuorumStoreInlineHybrid(b3, p3, m3)
+            },
+            (Payload::QuorumStoreInlineHybrid(b1, p1, m1), Payload::InQuorumStore(p2)) => {
+                // TODO: How to update m1?
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::QuorumStoreInlineHybrid(b1, p3, m1)
+            },
+            (Payload::QuorumStoreInlineHybrid(b1, p1, m1), Payload::InQuorumStoreWithLimit(p2)) => {
+                // TODO: What's the right logic here?
+                let m3 = if m1.is_none() {
+                    p2.max_txns_to_execute
+                } else if let Some(m2) = p2.max_txns_to_execute {
+                    Some(m1.unwrap() + m2)
+                } else {
+                    m1
+                };
+                let mut p3 = p1;
+                p3.extend(p2.proof_with_data);
+                Payload::QuorumStoreInlineHybrid(b1, p3, m3)
+            },
+            (Payload::InQuorumStore(p1), Payload::QuorumStoreInlineHybrid(b2, p2, m2)) => {
+                let mut p3 = p1;
+                p3.extend(p2);
+                Payload::QuorumStoreInlineHybrid(b2, p3, m2)
+            },
+            (Payload::InQuorumStoreWithLimit(p1), Payload::QuorumStoreInlineHybrid(b2, p2, m2)) => {
+                // TODO: What's the right logic here?
+                let m3 = if p1.max_txns_to_execute.is_none() {
```

Comment:
Same as above. Also see if its possible to refactor the logic into a common function.

Reply:
Refactored using a common function.